File: backend/app/schemas/workflow.py
"""
Agent 工作流 Pydantic Schemas
"""
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# --- 工作流节点定义 ---
class WorkflowNode(BaseModel):
    """工作流节点"""
    id: str = Field(..., description="节点唯一ID")
    type: str = Field(..., description="节点类型: llm_openai, llm_deepseek, llm_qwen, input, output, condition, tool_tts, tool_search")
    position: dict[str, float] = Field(default_factory=lambda: {"x": 0, "y": 0}, description="节点位置")
    data: dict[str, Any] = Field(default_factory=dict, description="节点配置数据")

    model_config = ConfigDict(extra="allow")


class WorkflowEdge(BaseModel):
    """工作流边（连接线）"""
    id: str = Field(..., description="边唯一ID")
    source: str = Field(..., description="源节点ID")
    target: str = Field(..., description="目标节点ID")
    sourceHandle: str | None = Field(None, description="源节点输出端口")
    targetHandle: str | None = Field(None, description="目标节点输入端口")
    label: str | None = Field(None, description="边标签（条件分支时使用）")


class WorkflowConfig(BaseModel):
    """完整工作流配置"""
    nodes: list[WorkflowNode] = Field(default_factory=list, description="节点列表")
    edges: list[WorkflowEdge] = Field(default_factory=list, description="边列表")


# --- API 请求/响应 ---
class WorkflowCreate(BaseModel):
    """创建工作流"""
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = None
    flow_data: WorkflowConfig | None = None


class WorkflowUpdate(BaseModel):
    """更新工作流"""
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = None
    flow_data: WorkflowConfig | None = None
    is_active: bool | None = None


class WorkflowResponse(BaseModel):
    """工作流响应"""
    id: int
    name: str
    description: str | None
    flow_data: dict[str, Any] | None
    is_active: bool
    created_by: int | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WorkflowExecuteRequest(BaseModel):
    """执行工作流请求"""
    workflow_id: int
    input_data: dict[str, Any] | None = None
    stream: bool = Field(default=True, description="是否流式输出")


class NodeResult(BaseModel):
    """节点执行结果"""
    node_id: str
    node_type: str
    status: str  # pending, running, completed, failed
    input: dict[str, Any] | None = None
    output: dict[str, Any] | None = None
    error: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None


class ExecutionResponse(BaseModel):
    """执行响应"""
    execution_id: int
    workflow_id: int
    status: str
    input_data: dict[str, Any] | None
    output_data: dict[str, Any] | None
    node_results: list[NodeResult] | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NodeDefinitionResponse(BaseModel):
    """节点定义响应"""
    id: int
    node_type: str
    name: str
    category: str
    description: str | None
    default_config: dict[str, Any] | None
    input_schema: dict[str, Any] | None
    output_schema: dict[str, Any] | None
    icon: str | None
    is_active: bool

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/workflow.ts
import request from '@/utils/request'

// 工作流节点类型
export interface WorkflowNodeData {
  label: string
  type: string
  config?: Record<string, unknown>
  [key: string]: unknown
}

export interface WorkflowNode {
  id: string
  type: string
  position: { x: number; y: number }
  data: WorkflowNodeData
}

export interface WorkflowEdge {
  id: string
  source: string
  target: string
  sourceHandle?: string
  targetHandle?: string
  label?: string
  animated?: boolean
  markerEnd?: string | { type: string; color?: string }
  type?: string
}

export interface WorkflowConfig {
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
}

export interface Workflow {
  id: number
  name: string
  description?: string
  flow_data?: WorkflowConfig
  engine_type?: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface NodeDefinition {
  id: number
  node_type: string
  name: string
  category: string
  description?: string
  default_config?: Record<string, unknown>
  input_schema?: Record<string, unknown>
  output_schema?: Record<string, unknown>
  icon?: string
}

export interface NodeResultItem {
  node_id: string
  status: 'pending' | 'running' | 'success' | 'failed'
  output?: unknown
  error?: string
  duration_ms?: number
}

export interface ExecutionResult {
  execution_id: number
  status: string
  output?: Record<string, unknown>
  node_results?: NodeResultItem[]
}

// API 函数
export const getWorkflows = async (params?: { skip?: number; limit?: number }) => {
  return request.get('/workflows/', { params })
}

export const getWorkflow = async (id: number) => {
  return request.get(`/workflows/${id}`)
}

export const createWorkflow = async (data: { name: string; description?: string; flow_data?: WorkflowConfig }) => {
  return request.post('/workflows/', data)
}

export const updateWorkflow = async (id: number, data: Partial<Workflow>) => {
  return request.put(`/workflows/${id}`, data)
}

export const deleteWorkflow = async (id: number) => {
  return request.delete(`/workflows/${id}`)
}

export const executeWorkflow = async (id: number, inputData?: Record<string, unknown>) => {
  return request.post(`/workflows/${id}/execute`, { workflow_id: id, input_data: inputData, stream: false })
}

// 流式执行
export const executeWorkflowStream = (id: number, inputData: Record<string, unknown>, onEvent: (_event: string, _data: ExecutionResult | NodeResultItem) => void) => {
  const eventSource = new EventSource(`/api/v1/workflows/${id}/execute?stream=true&input_data=${encodeURIComponent(JSON.stringify(inputData))}`)

  eventSource.onmessage = (e) => {
    try {
      const data = JSON.parse(e.data)
      onEvent(e.type, data)
    } catch {
      // Parse error silently
    }
  }

  eventSource.onerror = () => {
    eventSource.close()
  }

  return () => eventSource.close()
}

export const getNodeDefinitions = async () => {
  return request.get('/workflows/nodes/definitions')
}

export const getExecution = async (executionId: number) => {
  return request.get(`/workflows/executions/${executionId}`)
}
